Equip KubeArmor with Default Armors

daemon1024 opened the pull request:
Ref #595

daemon1024 committed:
introduce configurable default posture

KubeArmor didn't have a configurable default mode of operations. This commit introduces a configurable default posture as well changes in enforcement system to act accordingly.

Ref #595

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
redirect logs to alerts based on default posture

When KubeArmor is equipped with default posture block/audit each of the telemetry events generated needs to be an alert. This commit introduces changes to the policy matcher to update our logs to implicit block/audit alerts based on the configured default posture.

Ref #595

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

nyrahul commented on KubeArmor/config/config.go:
```diff
@@ -88,6 +101,10 @@ func readCmdLineParams() {
 	kvmAgentB := flag.Bool(ConfigKubearmorVM, false, "enabling KubeArmorVM")
 	k8sEnvB := flag.Bool(ConfigK8sEnv, true, "is k8s env?")
 
+	defaultFileArmor := flag.String(ConfigDefaultFileArmor, "allow", "configuring default enforcement action in global file context [audit,block]")
```
shouldn't the default armor be "block" in all the cases?
I am assuming the default armor comes into picture only when there is atleast one allow policy for the corresponding type (file/network/cap)?

nyrahul commented on KubeArmor/enforcer/appArmorEnforcer.go:
```diff
@@ -154,6 +159,22 @@ func NewAppArmorEnforcer(node tp.Node, logger *fd.Feeder) *AppArmorEnforcer {
 
+	ae.File = true
+	ae.Network = true
+	ae.Capability = true
```
Do we really need to keep the default armor in apparmor context ? Can we refer to cfg.DefaultArmor for all purposes?

nyrahul commented on KubeArmor/feeder/feeder.go:
```diff
@@ -536,7 +536,7 @@ func (fd *Feeder) PushLog(log tp.Log) {
 
 	// gRPC output
-	if log.Type == "MatchedPolicy" || log.Type == "MatchedHostPolicy" || log.Type == "MatchedNativePolicy" {
+	if log.Type == "MatchedPolicy" || log.Type == "MatchedHostPolicy" || log.Type == "MatchedNativePolicy" || log.Type == "DefaultArmor" {
```
I am not sure if I understand this. Are we having log.Type as 'DefaultArmor' inside the telemetry events?

nyrahul commented on KubeArmor/feeder/policyMatcher.go:
```diff
@@ -1008,18 +1008,58 @@ func (fd *Feeder) UpdateMatchedPolicy(log tp.Log) tp.Log {
-				return log
+			if log.Operation == "Process" {
+				if cfg.GlobalCfg.DefaultFileArmor == "block" {
+					log.Type = "DefaultArmor"
```
we should not change the `log.Type`.

nyrahul commented on KubeArmor/feeder/policyMatcher.go:
```diff
@@ -1008,18 +1008,58 @@ func (fd *Feeder) UpdateMatchedPolicy(log tp.Log) tp.Log {
+			if log.Operation == "Process" {
+				if cfg.GlobalCfg.DefaultFileArmor == "block" {
+					log.Type = "DefaultArmor"
+					log.Action = "ImplicitBlock"
```
Lets keep a global const for such string values.

nyrahul commented on KubeArmor/feeder/policyMatcher.go:
```diff
@@ -1008,18 +1008,58 @@ func (fd *Feeder) UpdateMatchedPolicy(log tp.Log) tp.Log {
+					return log
+				}
+			} else if log.Operation == "Network" {
+				if cfg.GlobalCfg.DefaultNetworkArmor == "block" {
```
This code looks redundant. Can we wrap this in function and have something like this?
```golang
func setLogFields(action string, log *types.Log) {
     if cfg.GlobalCfg.DefaultCapabilityArmor == "block" {
	(*log).Type = "DefaultArmor"
	(*log).Action = "ImplicitBlock"
     } else if cfg.GlobalCfg.DefaultCapabilityArmor == "audit" {
	  (*log).Type = "DefaultArmor"
	  (*log).Action = "ImplicitAudit"
     } else if (*log).CapabilitiesVisibilityEnabled {
	  (*log).Type = "ContainerLog"
     }
}

setLogFields(cfg.GlobalCfg.DefaultCapabilityArmor, &log)
```

daemon1024 committed:
consider default posture if fromsourceallowpolicy exists

This commit also fixes the bug where default deny didn't work if we only had fromSource based Allow Policies

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
check policy list to consider posture or not

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

daemon1024 committed:
consider default posture when atleast one allow policy OR from source allow policy

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

nyrahul reviewed (approved):
LGTM

daemon1024 commented on KubeArmor/enforcer/appArmorProfile.go:
```diff
@@ -1103,17 +1113,21 @@ func (ae *AppArmorEnforcer) GenerateProfileBody(securityPolicies []tp.SecurityPo
-		if file && len(processWhiteList) == 0 && len(fileWhiteList) == 0 {
+		if cfg.GlobalCfg.DefaultFilePosture == "block" && ((len(processWhiteList) > 0 || len(fileWhiteList) > 0) || !file) {
+		} else {
 			bodyFromSource = bodyFromSource + "    file,\n"
 		}
 
-		if network && len(networkWhiteList) == 0 {
+		if cfg.GlobalCfg.DefaultNetworkPosture == "block" && (len(networkWhiteList) > 0 || !network) {
+		} else {
 			bodyFromSource = bodyFromSource + "    network,\n"
 		}
 
-		if capability && len(capabilityWhiteList) == 0 {
+		if cfg.GlobalCfg.DefaultCapabilitiesPosture == "block" && (len(capabilityWhiteList) > 0 || !capability) {
+		} else {
 			bodyFromSource = bodyFromSource + "    capability,\n"
 		}
```
Updated to consider default posture in individual subprofiles using logic similar to what we use for generating profile head

nam-jaehyun commented on KubeArmor/config/config.go:
```diff
@@ -107,6 +124,10 @@ func readCmdLineParams() {
 	viper.Set(ConfigKubearmorVM, *kvmAgentB)
 	viper.Set(ConfigK8sEnv, *k8sEnvB)
 
+	viper.Set(ConfigDefaultFilePosture, *defaultFilePosture)
```
viper.SetDefault... let's update the PR on the latest code.

nam-jaehyun commented on tests/test-scenarios-github.sh:
```diff
@@ -266,7 +266,7 @@ function should_find_blocked_log() {
 
     if [[ $6 -eq 0 ]]; then
-        audit_log=$($CAT_LOG | grep -E "$1.*policyName.*\"$2\".*$match_type.*$3.*resource.*$4.*$5" | tail -n 1 | grep -v Passed)
+        audit_log=$($CAT_LOG | grep -E "$1.*policyName.*\"$2|DefaultPosture\".*$match_type.*$3.*resource.*$4.*$5" | tail -n 1 | grep -v Passed)
```
need to update all the three test-scenarios-X.sh together

nam-jaehyun reviewed (commented):
Let's update this PR on the latest code and merge it.
By the way, it would be better if we have a document that explains how the default posture works in specific cases.

nyrahul reviewed (approved):
LGTM

nam-jaehyun commented on KubeArmor/enforcer/appArmorProfile.go:
```diff
@@ -811,19 +812,27 @@ func (ae *AppArmorEnforcer) BlockedCapabilitiesMatchCapabilities(cap tp.Capabili
+	// AND
+	// -> Atleast one allow policy OR from source allow policy
+
+	if cfg.GlobalCfg.DefaultFilePosture == "block" && ((len(processWhiteList) > 0 || len(fileWhiteList) > 0) || !file) {
```
no statement for the if part?

daemon1024 replied:
If the condition matches, we want to block the access to resource, which essentially means we don't want want to add the resource to profile head, hence no statement

nam-jaehyun commented on KubeArmor/enforcer/appArmorProfile.go:
```diff
@@ -1103,17 +1113,21 @@ func (ae *AppArmorEnforcer) GenerateProfileBody(securityPolicies []tp.SecurityPo
 		}
 
-		if file && len(processWhiteList) == 0 && len(fileWhiteList) == 0 {
+		if cfg.GlobalCfg.DefaultFilePosture == "block" && ((len(processWhiteList) > 0 || len(fileWhiteList) > 0) || !file) {
```
no statement for the if part?

nam-jaehyun commented on KubeArmor/enforcer/appArmorProfile.go:
```diff
@@ -983,11 +996,6 @@ func (ae *AppArmorEnforcer) GenerateProfileBody(securityPolicies []tp.SecurityPo
-
-	profileHead := "  ## == PRE START == ##\n" + ae.GenerateProfileHead(processWhiteList, fileWhiteList, networkWhiteList, capabilityWhiteList) + "  ## == PRE END == ##\n\n"
-
 	// body
```
an empty line between 998 and 999

nam-jaehyun commented on tests/scenarios/github_test_12/cmd1:
```diff
@@ -0,0 +1,7 @@
+source: ubuntu-1-deployment
+cmd: curl 142.250.193.46
```
specific IP? why do we need to use it?
is it to avoid a UDP connection for DNS resolution?

daemon1024 replied:
> is it to avoid a UDP connection for DNS resolution?

Yes

nam-jaehyun commented on tests/scenarios/github_test_12/cmd2:
```diff
@@ -0,0 +1,7 @@
+---
+operation: Network
+condition: SOCK_DGRAM
+action: Block
```
the policy action is "Allow". why is it "Block"?

daemon1024 replied:
The policy action is allow but default posture is Block, this will match the default Posture.

nam-jaehyun commented on tests/scenarios/github_test_12/cmd3:
```diff
@@ -0,0 +1,7 @@
+---
+operation: Network
+condition: SOCK_STREAM
+action: Block
```
why "Block"?

daemon1024 replied:
The policy action is allow but default posture is Block, this will match the default Posture.

nam-jaehyun commented on tests/scenarios/github_test_12/cmd3:
```diff
@@ -0,0 +1,7 @@
+source: ubuntu-1-deployment
+cmd: wget --tries=1 142.250.193.46
```
what is the difference between cmd1 and cmd3?

daemon1024 replied:
We have a policy for allowing tcp for curl, this tries to access tcp using wget which should be blocked due to our default posture

nam-jaehyun commented on tests/scenarios/github_test_12/ksp-ubuntu-1-net-tcp-from-source-allow.yaml:
```diff
@@ -0,0 +1,17 @@
+    matchProtocols:
+    - protocol: tcp
+      fromSource:
+              - path: /usr/bin/curl
```
tab -> space

nam-jaehyun commented on KubeArmor/feeder/policyMatcher.go:
```diff
@@ -873,6 +900,9 @@ func (fd *Feeder) UpdateMatchedPolicy(log tp.Log) tp.Log {
 							continue
 						}
 					}
+					if secPolicy.Action == "Allow" {
```
an empty line between 902 and 903

nam-jaehyun reviewed (commented):
Added some comments. Please check them.

daemon1024 committed:
add default posture block test case based fromsourceallow

Modified test script to accomodate DefaultPosture logs

Signed-off-by: daemon1024 <barun.acharya@accuknox.com>

nam-jaehyun reviewed (approved):
Let's merge it.
